floodsub.go

```diff
@@ -324,6 +359,94 @@ func msgID(pmsg *pb.Message) string {
 	return string(pmsg.GetFrom()) + string(pmsg.GetSeqno())
 }
 
+// pushMsg pushes a message to a number of subscriptions, performing validation
+// as necessary
+func (p *PubSub) pushMsg(subs []*Subscription, src peer.ID, msg *Message) {
+	// we perform validation if _any_ of the subscriptions has a validator
+	// because the message is sent once for all topics
+	needval := false
+	for _, sub := range subs {
+		if sub.validate != nil {
+			needval = true
+			break
+		}
+	}
+
+	if needval {
+		// validation is asynchronous and globally throttled with the throttleValidate semaphore
+		select {
+		case p.validateThrottle <- struct{}{}:
+			go func() {
+				p.validate(subs, src, msg)
+				<-p.validateThrottle
+			}()
+		default:
+			log.Warningf("message validation throttled; dropping message from %s", src)
+		}
+		return
+	}
+
+	sreq := &sendReq{from: src, msg: msg}
+	select {
+	case p.sendMsg <- sreq:
+	default:
+		go func() {
+			p.sendMsg <- sreq
+		}()
+	}
+}
+
+// validate performs validation and only sends the message if all validators succeed
+func (p *PubSub) validate(subs []*Subscription, src peer.ID, msg *Message) {
+	ctx, cancel := context.WithCancel(p.ctx)
+	defer cancel()
+
+	results := make([]chan bool, 0, len(subs))
+	throttle := false
+
+loop:
+	for _, sub := range subs {
+		if sub.validate == nil {
+			continue
+		}
+
+		rch := make(chan bool, 1)
```

Comment:
I'd just use a single channel and count results from it. Allocating channels isn't *that* expensive but it's certainly not free.

Reply:
fair enough; it was _really_ convenient to write this way though :)

Reply:
alright, implemented!